# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

In [30]:
! git clone https://www.github.com/DS3001/assignment2


Cloning into 'assignment2'...
remote: Enumerating objects: 36, done.
remote: Counting objects: 100% (12/12), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 36 (delta 8), reused 5 (delta 5), pack-reused 24
Receiving objects: 100% (36/36), 5.47 MiB | 20.58 MiB/s, done.
Resolving deltas: 100% (9/9), done.


**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

###Q1

1)

The abstract states that this paper is about the development of a tidying method. Data tidying is the process of giving data a specific strucutre which is then easy to perform analysis and visualization on.

2)

According to the introduction the Tidy data standard is used to create a go to method for data cleaning such that new instances do not need to be overly creative with how to organize the data. It is based on several mathmatical principles and research

3)

The sentance  "Like families, tidy datasets are all alike but every messy dataset is messy in its own way," refers to the standardization that comes form tidy data. Tidy data will all have similar attributes. However messy data can be messy in nearly infinite ways. This means it is impossible to predict just what makes messy data messy.

The senatnce  "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general," means that it is generally intutitve whether somethign is an observation or a varible likely based on whether it is in a row or column. Also, depending on the situation for the data the recording could be either one.

4)

Values = is the actual information (strings or integers) that are in the data set

variables = measures an attribute across all units

observation = contains all attributes on same unit

5)

This definition comes from Codd's 3rd normal form. It states: all varibales form a column, each observation form a row, each type of observation unit forms a table.

6)

The five most common issues are:

• Column headers are values, not variable names.

• Multiple variables are stored in one column.

• Variables are stored in both rows and columns.

• Multiple types of observational units are stored in the same table.

• A single observational unit is stored in multiple tables.

Table four needs to be melted. This is the process of turning rows into columns.

7)

Table 11 is messy because there are variables in columns, across coumns, and in rows. Table 12 fixes this by first matching ID,year, and month to the tmin and tmax variables, then making each row its own observation by placing varibles tmin and tmax in columns.

8)

The chicken and egg problem is whether or not the tools for tidy data analysis or the tidy data itself is the orignial start to the self reinforcing loop of needing tidy data. She hopes that this will only be the start and that there will be more development in Tidy data to come.


**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [18]:
import numpy as np
import pandas as pd
from pandas.core.base import value_counts

######1

In [73]:

air = pd.read_csv("/content/assignment2/data/airbnb_hw.csv")
print(air.dtypes)
air.head()


Host Id                         int64
Host Since                     object
Name                           object
Neighbourhood                  object
Property Type                  object
Review Scores Rating (bin)    float64
Room Type                      object
Zipcode                       float64
Beds                          float64
Number of Records               int64
Number Of Reviews               int64
Price                          object
Review Scores Rating          float64
dtype: object


Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  
0                  1                  0   145                   NaN  
1                  1                  1    37                   NaN  
2                  1                  1    28                   NaN  
3                  1                  0   199                   NaN  
4                  1                 39   549                  96.0

In [74]:
air["Price"] = air["Price"].str.replace(",", "") #This replaces the , with empty space. this is important to allow for coercion later
air["Price"] = air["Price"].astype(int) # this makes the datatype integer. this allows us to do numeric analysis


In [75]:
air["Price"] = air["Price"].astype(int)
air["Price_nan"] = air["Price"].isnull() #this creates a missing value dummy
value_counts(air["Price_nan"])

False    30478
Name: Price_nan, dtype: int64

In [76]:
print(air.dtypes)


Host Id                         int64
Host Since                     object
Name                           object
Neighbourhood                  object
Property Type                  object
Review Scores Rating (bin)    float64
Room Type                      object
Zipcode                       float64
Beds                          float64
Number of Records               int64
Number Of Reviews               int64
Price                           int64
Review Scores Rating          float64
Price_nan                        bool
dtype: object


####2

In [ ]:
df2 = pd.read_csv("/content/assignment2/data/sharks.csv",low_memory=False)
print(df2["Type"].unique())
print(df2["Type"].value_counts())
#print(df2.dtypes)
df2.head()

In [ ]:
from pandas._libs.hashtable import value_count
df2["Type"] = df2["Type"].replace(["Boat","Boating","Boatomg"],"Watercraft") #multiple categorical responses could be realisitcally considered "boats of some sort"
df2["Type"] = df2["Type"].replace(["Questionable","Unverfified","Under investigation","Unverified"], "Unconfirmed") # the low response rate suggests that combination is unproblematic. All have an element of uncertainty.
print(value_counts(df2["Type"]))




3

In [19]:
url="http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv"
df = pd.read_csv(url,low_memory=False)

In [ ]:
df["WhetherDefendantWasReleasedPretrial"] = df["WhetherDefendantWasReleasedPretrial"].replace(9,np.nan)
df=df.rename(columns= {"WhetherDefendantWasReleasedPretrial":"released"})
print(value_counts(df["released"]))

1.0    19154
0.0     3801
Name: released, dtype: int64


4

In [ ]:
 #Codebook states that all charges led to pending, deferred dropped. Thus for analysis purposes these charges led to 0 years jailtime. Potential for error due to pending charges, but this could be a flaw
 #in the recording since it is included with other states that could be thought of as no jail time.
df.loc[(df["SentenceTypeAllChargesAtConvictionInContactEvent"]==4), "ImposedSentenceAllChargeInContactEvent" ] = 0
#this state, 9,  suggests the lack of information so operates as a nan value.
df.loc[(df["SentenceTypeAllChargesAtConvictionInContactEvent"]==9), "ImposedSentenceAllChargeInContactEvent" ] = np.nan
df["ImposedSentenceAllChargeInContactEvent"] = pd.to_numeric(df["ImposedSentenceAllChargeInContactEvent"],errors='coerce')
df["ImposedSentenceAllChargeInContactEvent"] = df["ImposedSentenceAllChargeInContactEvent"].apply(np.ceil) #for analysis rounds to closest integer above float. Helps analysis but doesnt round down time (ie avoids weekend sentances being 0)
df["imposed_nan"]=df["ImposedSentenceAllChargeInContactEvent"].isnull()
print(value_counts(df["imposed_nan"]))
print(value_counts(df["ImposedSentenceAllChargeInContactEvent"]))
pd.crosstab(df['ImposedSentenceAllChargeInContactEvent'], df['SentenceTypeAllChargesAtConvictionInContactEvent'])


**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

In [71]:
air = pd.read_csv("/content/assignment2/data/airbnb_hw.csv")


2. The dimesions of this data 30478x13. This is 13 variables and 30478 observations. Variables are displayed by air.dtypes below.

In [ ]:
air.head()
#print(air.dtypes)

In [ ]:
#3
pd.crosstab(air["Room Type"],air["Property Type"])

3. cont'd
Overall trends show that apartments are the most common type of property and that entire home/apt is the most common room type. The properties that a private room is more common for are B&B, House,Form,Other.

In [85]:
import seaborn as sns

In [ ]:
#4
air["Price"].describe()



In [ ]:
air.boxplot("Price")
air.hist("Price",bins=100)

In [ ]:
sns.kdeplot(air["Price"])

There are many outliers and the data is badly scewed.

In [93]:
air["price_log"]=np.log(air["Price"])

In [ ]:
air["price_log"].describe()


In [ ]:
air.boxplot("price_log")
air.hist("price_log",bins=100)

In [ ]:
sns.kdeplot(air["price_log"])

In [ ]:
#5
sns.scatterplot(x=air["price_log"],y=air["Beds"])


In [ ]:
air.groupby("Beds")["Price"].describe()
#This breakdown shows that there are some bed numbers with very high SD. notably 5, 8, and 12. This could mean that these types of houses hold outlier data. Also the appearance of listings decreases directly
#the number of beds. The mean generally increases with some major outliers (5,8,12) which are likely affected by outliers.

Make a scatterplot of price_log and Beds, but color the graph by Room Type and Property Type. What patterns do you see? Compute a description of Price conditional on Room Type and Property Type. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?

In [ ]:
#6
sns.scatterplot(x=air["price_log"],y=air["Beds"],hue=air["Room Type"],style= air["Property Type"])
#This showsw that most of the ourtliers in price are entire rooms, and beyond that many are house with a few apartments.
#Private rooms live at a log price between 3.5 and 4.5 mostly
#shared rooms are the least common and exist at the lowest end of the log price scale.


In [ ]:
air.groupby("Property Type")["Price"].describe()
#Other has the highest price and SD, this is likely an unusual property that doesnt match the describtion.
# after that it is VIlla
# the median seems more accurate since there are definetly outliers

In [ ]:
air.groupby("Room Type")["Price"].describe()
#the highest mean and SD is entire home/apt
# the median seems more accurate since there are definetly outliers

In [ ]:
#7
sns.jointplot(x=air["price_log"],y=air["Beds"],kind="hex")
#The data really only exists between 3 and 6 log price and 1 and 3 bedroom. Thus there is an incredible effect of outliers that should be removed for analysis.

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.